Change metadata service helper to work with any query parameters

davidwtbuxton opened the pull request:
Part of #579 

This helper is used with '?recursive=true' in one place, and can now be used by
IDTokenCredentials for requests with query parameters to the metadata identity
end-point.

This change will allow making requests to the token end-point with '?scopes=..'
query parameters.

busunkim96 commented:
Hi @davidwtbuxton,

It looks like there are two things that aren't passing in the `Kokoro` job:


- cover
    ```
    .nox/cover/lib/python3.7/site-packages/google/auth/compute_engine/_metadata.py        70      0     16      1    99%   
    142->145
    ```
- lint
Install nox `pip install nox` and run the autoformat session `nox -s blacken`. If this is troublesome I can also push to yoru branch.
    ```
    black --check google tests noxfile.py setup.py docs/conf.py
    would reformat /tmpfs/src/github/google-auth-library-python/google/auth/compute_engine/_metadata.py
    would reformat /tmpfs/src/github/google-auth-library-python/google/auth/compute_engine/credentials.py
    All done! 💥 💔 💥
    2 files would be reformatted, 70 files would be left unchanged.
    ```

davidwtbuxton commented:
Nuts. I don't understand why that branch isn't covered by the tests. I will run black to satisfy the linter, and dig deeper on the coverage issue.

tseaver commented on google/auth/compute_engine/_metadata.py:
```diff
@@ -133,12 +137,17 @@ def get(request, path, root=_METADATA_ROOT, recursive=False, retry_count=5):
+    if params is not None:
+        url = _helpers.update_query(base_url, params)
+    else:
+        url = base_url
```
This would be clearer if it just used `query_params` as before, defaulting it, e.g.
```python
    base_url = urlparse.urljoin(root, path)
    query_params = {} if params is None else params

    if recursive:
        query_params["recursive"] = "true"

    url = _helpers.update_query(base_url, params)
```

tseaver reviewed (changes requested):
Please ensure that unit tests pass on all supported versions of Python.   The best path is to run `nox`, after ensuring that `python3.6`, `python3.7`, and `python3.8` are on your path.

tseaver commented:
@davidwtbuxton You are also going to need to test with `python2.7` and `python3.5', exercised via `nox -s unit_prev_versions`.

davidwtbuxton committed:
feat: Change metadata service helper to work with any query parameters

This helper is used with '?recursive=true' in one place, and can now be used by
IDTokenCredentials for requests with query parameters to the metadata identity
end-point.

This change will allow making requests to the token end-point with '?scopes=..'
query parameters.

davidwtbuxton commented:
@tseaver I've updated the code with your suggestion for simplifying the params logic, thank you.

Kokoro failed because of the unrelated changes in commit 7e1525822d51bd9ce7dffca42d71313e6e776fcd , and tests are also broken in master. I've rebased changes in this branch on the top of release v1.21.3 (from 23 September 2020) and all tests pass now, but of course this feature/579-app-engine-scopes branch is out-of-date with the head of master.

The `unit_prev_versions` session isn't in noxfile, but I did run `nox -s unit` with Python 2.7 and Python 3.5 -> 3.8, all tests passed.

tseaver commented:
@davidwtbuxton Tests are failing on `master` for `unit-3.6` and `unit-3.7`  (#629), due to https://github.com/googleapis/google-auth-library-python/issues/629.

tseaver commented:
@davidwtbuxton The `unit_prev_versions` session is definitely [present on `master`](https://github.com/googleapis/google-auth-library-python/blob/6407258956ec42e3b722418cb7f366e5ae9272ec/noxfile.py#L93-L99).

davidwtbuxton commented:
@tseaver I've updated this branch with the latest changes from master, including the fixes from #630. Thank you, David

busunkim96 commented:
@silvolu is taking this change alright, or should we be waiting for the PR to add a different credential type?

silvolu commented:
Yeah we need this to pass scopes to the metadata server, thanks @davidwtbuxton!
I'm working on a PR to change Compute Credentials to accept user defined scopes, we decided to be consistent with the existing implementation on Go and Node where we allow developers to specify scopes, and then depending on where they are running they get a token with the scopes they requested (e.g. GKE) or with the predefined scopes (e.g. GCE).

davidwtbuxton commented:
@silvolu sounds like #633
